lib/resque/scheduler/lua.rb

```diff
@@ -1,46 +1,51 @@
+require "digest"
+
 module Resque
   module Scheduler
     module Lua
       extend self
 
-      LUA_SCRIPTS_PATH = File.join(File.dirname(__FILE__), 'lua')
-
       def zpop(key, min, max, offset, count)
-        evalsha(:zpop, [key], [min, max, offset, count])
+        ZPOP.call(Resque.redis, [key], [min, max, offset, count])
       end
 
       def deleq(key, expected_value)
-        evalsha(:deleq, [key], [expected_value])
+        DELEQ.call(Resque.redis, [key], [expected_value])
       end
 
       def locked(lock_key, token, timeout)
-        evalsha(:locked, [lock_key], [token, timeout])
+        LOCKED.call(Resque.redis, [lock_key], [token, timeout])
       end
 
       private
 
-      def get_sha(script, refresh = false)
-        sha_store[script] = nil if refresh
-        sha_store[script] ||= Resque.redis.script(:load, load_lua(script))
-      end
+      LUA_SCRIPTS_PATH = File.join(File.dirname(__FILE__), 'lua')
 
-      def evalsha(script, keys, argv, refresh: false)
-        Resque.redis.evalsha(get_sha(script, refresh), keys: keys, argv: argv)
-      rescue Redis::CommandError => e
-        if e.message =~ /NOSCRIPT/
-          refresh = true
-          retry
+      Script = Struct.new(:body) do
+        def self.from_file(filename)
+          new(File.read(LUA_SCRIPTS_PATH + "/#{filename}.lua"))
         end
-        raise
-      end
 
-      def load_lua(filename)
-        File.read(LUA_SCRIPTS_PATH + "/#{filename}.lua")
-      end
+        def call(conn, keys, argv)
+          conn.evalsha(sha1, keys: keys, argv: argv)
+        rescue Redis::CommandError => ex
+          if ex.message =~ /\ANOSCRIPT/
+            conn.eval(body, keys: keys, argv: argv)
+          else
+            raise
+          end
+        end
+
+        private
 
-      def sha_store
-        @sha_store ||= {}
+        def sha1
+          @sha ||= Digest::SHA1.hexdigest(body)
+        end
       end
+
+      ZPOP = Script.from_file("zpop")
```

Comment:
ZPOP is a new operation in Redis 5. We can remove this script once we upgrade :)